modules/devices/steamdeck/fde.nix

```diff
@@ -0,0 +1,105 @@
+
+              Plymouth and the console work by default.
+            '';
+          };
+          debug = lib.mkOption {
+            default = false;
+            type = lib.types.bool;
+            description = ''
+              Enable debug output for deckbd.
+
+              This will log keycodes to the journal.
+            '';
+          };
+        };
+      };
+    };
+  };
+
+  config = lib.mkMerge [
+    (lib.mkIf cfg.fde.deckbd.enable {
+      assertions = [
+        {
+          assertion = cfg.enableDefaultStage1Modules;
+          message = ''
+            deckbd cannot be used without the required kernel modules.
+
+            jovian.devices.steamdeck.enableDefaultStage1Modules must be enabled.
+          '';
+        }
+        {
+          assertion = config.boot.initrd.systemd.enable;
+          message = ''
+            deckbd requires systemd-in-initrd.
+
+            boot.initrd.systemd.enable must be set to true. Note that this may impact other NixOS features.
+          '';
+        }
+      ];
+
+      jovian.devices.steamdeck.fde.deckbd.bindsTo = [
+        "systemd-ask-password-console.service"
+        (lib.mkIf config.boot.plymouth.enable config.systemd.services.systemd-ask-password-plymouth.name)
+        # Note: Agents that have optional keyboard input, such as unl0kr, should not be added here. The user may not want it.
+      ];
+
+      boot.initrd = {
+        kernelModules = [ "uinput" ];
+
+        services.udev.packages = [
+          (pkgs.writeTextDir "lib/udev/rules.d/70-steam-deck-controller.rules" ''
+            ACTION=="add", SUBSYSTEM=="input", ENV{ID_VENDOR_ID}=="28de", ENV{ID_MODEL_ID}=="1205", ENV{PRODUCT}=="?*", ENV{ID_INPUT_JOYSTICK}=="1", TAG+="systemd", ENV{SYSTEMD_ALIAS}+="/sys/class/input/input_steam_deck_controller"
+          '')
+        ];
+
+        systemd = {
+          storePaths = with pkgs; [ deckbd ]; # https://github.com/NixOS/nixpkgs/issues/309316
+          services.deckbd = {
+            description = "Steam Deck controller keymapper";
+            unitConfig.DefaultDependencies = false;
+            wantedBy = cfg.fde.deckbd.bindsTo;
+            before = cfg.fde.deckbd.bindsTo;
+            inherit (cfg.fde.deckbd) bindsTo;
+            requires = [ "sys-class-input-input_steam_deck_controller.device" ];
+            after = [ "sys-class-input-input_steam_deck_controller.device" ];
+            path = with pkgs; [ deckbd ];
+            environment.G_MESSAGES_DEBUG = lib.mkIf cfg.fde.deckbd.debug "all";
+            # Note: The lizard_mode parameter does not work as intended. The
+            # Steam Deck must also be manually put into gamepad mode by holding
+            # the menu button (≡) for a second.
+            preStart = "echo 0 > /sys/module/hid_steam/parameters/lizard_mode";
+            script = "deckbd";
+            postStop = "echo 1 > /sys/module/hid_steam/parameters/lizard_mode";
```

Comment:
I've dived into trying to understand why it wouldn't work "as expected".

First, here's what the `lizard_mode` parameter is documented to do:

> This driver will disable the lizard mode when the input device is opened
> and re-enable it when the input device is closed, so as not to break user
> mode behaviour. The lizard_mode parameter can be used to change that.

 - https://github.com/Jovian-Experiments/linux/blob/b9bc2d44db5b6ea54333d20a02de16780e41acb9/drivers/hid/hid-steam.c#L18-L20

This comment has aged poorly... That's only true for the Steam Deck controller... But still has interesting hints as to what is happening.

So, here's the reason why it "doesn't work" for the Steam Deck controller:

 - https://github.com/Jovian-Experiments/linux/blob/b9bc2d44db5b6ea54333d20a02de16780e41acb9/drivers/hid/hid-steam.c#L597-L608

... Except that this is only about automatically enabling/disabling lizard mode ***on open***. And that's why things gets kinda confusing fast.

When the parameter is changed *live*, this happens:

 - https://github.com/Jovian-Experiments/linux/blob/b9bc2d44db5b6ea54333d20a02de16780e41acb9/drivers/hid/hid-steam.c#L1818-L1821

See how there's no check for the hardware in use? So, as long as the device is not *opened*, the switch will happen... ***EXCEPT*** it's not that easy...

If the device has been “forced” into *gamepad mode*, attempting to set it to lizard mode *won't work*.

 - https://github.com/Jovian-Experiments/linux/blob/b9bc2d44db5b6ea54333d20a02de16780e41acb9/drivers/hid/hid-steam.c#L557-L560

In turn this is what the (≡) button does; it flips that flag:

 - https://github.com/Jovian-Experiments/linux/blob/b9bc2d44db5b6ea54333d20a02de16780e41acb9/drivers/hid/hid-steam.c#L1052

(Though also notice how if `lizard_mode` is disabled, it will not be put into `lizard_mode`...)

* * *

So, to recap, ***for a steam deck***, when the device is *registered to the driver*, if *it was not already opened*, it will be out into the mode set by the kernel module parameter:

 - https://github.com/Jovian-Experiments/linux/blob/b9bc2d44db5b6ea54333d20a02de16780e41acb9/drivers/hid/hid-steam.c#L990-L991

***If*** the device is not opened, *changing the kernel module parameter* will change the mode of the controller...

 - https://github.com/Jovian-Experiments/linux/blob/b9bc2d44db5b6ea54333d20a02de16780e41acb9/drivers/hid/hid-steam.c#L1818-L1821

... unless it's been “put” into gamepad mode (via the ≡ button)

 - https://github.com/Jovian-Experiments/linux/blob/b9bc2d44db5b6ea54333d20a02de16780e41acb9/drivers/hid/hid-steam.c#L557-L558

And the last situation where the `lizard_mode` parameter takes effect on the Steam Deck is when the `hid_ll_driver` interface for it is closed (which is of no importance for us here):

 - https://github.com/Jovian-Experiments/linux/blob/b9bc2d44db5b6ea54333d20a02de16780e41acb9/drivers/hid/hid-steam.c#L1138

* * *

Did you get all that? :) (I'm still somewhat confused.)